### PR TITLE
Randomized allocation sampling

### PR DESCRIPTION
**Pull request by noahfalk**
This feature allows profilers to do allocation profiling based off randomized samples. It has better theoretical and empirically observed accuracy than our current allocation profiling approaches while also maintaining low performance overhead. It is designed for use in production profiling scenarios. For more information about usage and implementation, see the included doc docs/design/features/RandomizedAllocationSampling.md

[Updated 10/30]
This PR supersedes https://github.com/dotnet/runtime/pull/100356. Currently it includes 2 commits, the first has the (mostly) rebased changes from July and the 2nd addressed a few things I missed in the rebase and followed up on issues flagged in July. 

Most of the testing happened during July but I did re-run some of the functional tests as a smoke test to ensure nothing broke in the few recent changes. We'll also get updated CI regression testing and I'm coordinating with @chrisnas to have him validate it works in a real profiler.

### CODE REVIEW

File: src/coreclr/nativeaot/Runtime/GCHelpers.cpp
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/GCHelpers.cpp:627]: Do we need to account for unused sampling quota from before the last combined limit reset?

The scenario that I am thinking about: One thread allocates a lot and triggers the GC a lot, second thread allocates a little. The second thread allocates a little. The second thread should get its fair share of samples, but I do not think it is going to the case since its sampling quota will be reset by GC running before it gets a chance to report a sample most of the time.

Comment (noahfalk) [reply src/coreclr/nativeaot/Runtime/GCHelpers.cpp:627]: > Do we need to account for unused sampling quota from before the last combined limit reset?

To the best of my understanding, discarding the remainder of sampling quota and regenerating the limit does not cause any problem for statistical accuracy as long as the decision to discard is independent of the sample limit value. There is some additional discussion of this late in the feature doc: https://github.com/dotnet/runtime/pull/104955/files#diff-f0760b2f7659c7433b5313f11d30bf17f0061252058af8511ad117bfee99ceb3R310

> but I do not think it is going to the case with this implementation since its sampling quota will be reset by GC running before it gets a chance to report a sample most of the time.

The sampling quota is generated in such a way that every allocated byte has a 1/102,400 chance of being sampled. For example if thread A has only allocated 20 bytes between each GC then it will have  ~20/102,400 chance of sampling at least one byte during that time. That is a small chance, but it is the proportionally correct chance based on the amount of allocating this thread performed.

**Comment by noahfalk** [comment]
Functional testing found and fixed an off-by-one error in the RNG code but otherwise things looked fine. I also resynced this PR on top of the latest changes in https://github.com/dotnet/runtime/pull/104849 and https://github.com/dotnet/runtime/pull/104851. The last commit, now number 10, remains the interesting one.

I also did some performance testing using GCPerfSim as an allocation benchmark. My default configuration was 4 threads, workstation mode, 500GB of allocations entirely with small objects and no survival. It is intended to put maximum stress on the allocation code paths. GCPerfSim command line: `-tc 4 -tagb 500 -tlgb 0.05 -lohar 0 -lohsr 100000-2000000 -sohsi 0 -lohsi 0 -pohsi 0 -sohpi 0 -lohpi 0 -sohfi 0 -lohfi 0 -pohfi 0 -allocType reference -testKind time`. I also ran a few variations that added modest amounts of survival and LOH allocations that are a bit more realistic, though still extremely allocation heavy.

### Benchmarks - No Tracing enabled

| Scenario                                    |    Baseline time   |    PR  time   |
|---------------------------------------------|--------------------|---------------|
| Default                                     |     20.8           |    21.4       |
| Default + `sohsi 100`                       |     40.9           |    41.8       |
| Default + `sohsi 100 -lohar 10`             |     42.8           |    43.9       |
| Default + `sohsi 100 -lohar 10 -lohsi 200`  |     43.6           |    44.6       |

### Benchmarks - Tracing AllocSampling+GC keywords, verbose level

| Scenario                                    |    Baseline time   |    PR  time   |
|---------------------------------------------|--------------------|---------------|
| Default                                     |     22             |    23.1       |
| Default + `sohsi 100`                       |     41.6           |    43.0       |
| Default + `sohsi 100 -lohar 10`             |     44.7           |    44.5       |
| Default + `sohsi 100 -lohar 10 -lohsi 200`  |     44.2           |    45.0       |

Overall it looks like around ~0.9 additional seconds for the PR to do 500GB of allocations. On a tight microbenchmark its noticeable and then as other GC or non-allocation costs increase it becomes relatively less noticeable. I'm investigating to see if it can be improved at all.

File: src/coreclr/nativeaot/Runtime/thread.inl
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/thread.inl:73]:
```suggestion
    if(value == UINT32_MAX)
    {
        value--;
    }
    return value * (1.0/UINT32_MAX);
```
```suggestion
    return value * (1.0/(UINT32_MAX+1));
```
Mapping UINT32_MAX to UINT32_MAX-1 would make UINT32_MAX-1 two times more likely than other values.

Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/thread.inl:73]:
```suggestion
    return value * (1.0/(UINT32_MAX + 1.0));
```
Mapping UINT32_MAX  to UINT32_MAX-1 would make UINT32_MAX-1 two times more likely than other values.

Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/thread.inl:73]:
```suggestion
    return value * (1.0/(UINT32_MAX + 1.0));
```
Mapping UINT32_MAX  to UINT32_MAX-1 would make UINT32_MAX-1 two times more likely than other values.

Comment (noahfalk) [reply src/coreclr/nativeaot/Runtime/thread.inl:73]: Agreed, it was a copy of what CLRRandom did knowing that the duplicate wouldn't be consequential to the statistical results. Changing the divisor by 1 will mean the subset of random values that do generate sample limits below 8KB (UINT32_MAX isn't one of them) will be ~0.000002 bytes smaller than before. I'm fine to make the change though.

**Comment by noahfalk** [comment]
Continued perf investigation+testing has cast my previous results into doubt. After lots of searching for what could have caused the regression, my best explanation is that it actually had nothing to do with the source changes in this PR and instead it is either non-determinism in the build process or some user error on my part. I reached that conclusion by doing the following:

I've had a folder on my machine C:\git\runtime3 that throughout the entire process has been synced here:
```
commit 42b2b19e883f06af5771b5d85b26af263c62e781 (HEAD)
Author: Matous Kozak <55735845+matouskozak@users.noreply.github.com>
Date:   Fri Jul 12 09:42:55 2024 +0200
```
This folder has no changes from any of my PRs in it and I've been using the build here for all the baseline measurements. Then I executed the following changes:
```
move artifacts -> artifacts_backup
build.cmd clr+libs -c release
src\tests\build.cmd generatelayoutonly Release
copy artifacts_backup -> artifacts_backup_2
```

I can consistently reproduce the same magnitude perf regression using the coreclr built in the artifacts directory, but the regression doesn't appear using the build in the backup or backup_2 directory. I've done many runs on each binary switching between them in a semi-randomized ordering trying to ensure that the results for each binary are repeatable and robust relative to background noise on the machine.

Beyond that I've also got many other builds that include different subsets of the change but there is no clear relationship between the source and the perf results. During one period I progressively added functionality starting from the baseline without triggering the regression to occur, then during another period I was progressively removing functionality from the final PR and the regression would always occur. Even deleting the entirety of the source changes in that folder and syncing it back to the baseline didn't eliminate the perf overhead. Every build was done in a new folder starting without an artifacts folder to remove opportunity for incremental build problems to play a role.

The only explanations I have that make sense to me are either: (a) non-deterministic builds are giving bi-modal perf results for the same input source code or (b) I am making some other error in my testing methodology repeatedly

I'm going to see if I can get another machine to repeat some of the original experiments but at the moment I no longer have any evidence the PR is causing a regression.

**Comment by noahfalk** [comment]
@jkotas @MichalStrehovsky - Functional and perf testing both looked good now, all outstanding comments on the PRs have been addressed, and CI is green. From my perspective this is ready to be merged unless any further review is planned?

I could check in https://github.com/dotnet/runtime/pull/104849, https://github.com/dotnet/runtime/pull/104851, then this PR in sequence but I'm not sure that gives any advantage over just checking in this PR alone and closing 104849 and 104851 as no longer needed.

**Comment by jkotas** [comment]
- GC stress run identified a functional issue in https://github.com/dotnet/runtime/pull/104849: https://github.com/dotnet/runtime/pull/104849#issuecomment-2254934408

- https://github.com/dotnet/runtime/pull/104851 is still running native AOT tests. Also, there is a question for Mike about the data contract.

File: src/tests/tracing/eventpipe/randomizedallocationsampling/manual/README.md
Comment (jkotas) [inline src/tests/tracing/eventpipe/randomizedallocationsampling/manual/README.md:1]: What prevents these tests from being automated?

**Review by jkotas — COMMENTED** [review]
> this PR in sequence but I'm not sure that gives any advantage over just checking in this PR alone 

I think it still makes sense. I am not familiar enough with tracing to tell whether it is hooked up correctly. It would be best to for somebody familar with the tracing to review that part. It will be much easier if the delta does not show the other changes.

Also, this is non-trivial feature (a few thousand lines), with risk of introducing regressions (as demonstrated by the GC stress crash). Given that we are feature complete for .NET 9, should this get an approval from Jeff or tactics before merging for .NET 9?

File: src/tests/tracing/eventpipe/randomizedallocationsampling/manual/README.md
Comment (noahfalk) [reply src/tests/tracing/eventpipe/randomizedallocationsampling/manual/README.md:1]: The test as currently designed does large numbers of allocations with different types of objects and then prints statistical distributions to the console. Nothing fundamental prevents it from being converted to an automated test but the main factors that discourage it are:
- its a slow test even on a fast machine - it does large amounts of allocations to gather many statistical distributions
- it would require effort to convert it and ensure we've reasonably calibrated the sensitivity for the randomized results the test generates.

Historically our automated testing on tracing events validate that the events are generated and check some fields for reasonable data, but don't go into great detail. That level of testing has had good cost-benefit tradeoff in the past so we repeated it here.

**Comment by noahfalk** [comment]
> Given that we are feature complete for .NET 9, should this get an approval from Jeff or tactics before merging for .NET 9?

Yes, I was assuming that would be part of the process.

File: src/coreclr/vm/gchelpers.cpp
Comment (noahfalk) [reply src/coreclr/vm/gchelpers.cpp]: Fixed, thanks!

**Comment by noahfalk** [comment]
@brianrob @chrisnas - I wanted to give a heads up that I removed the HeapIndex parameter on the AllocationSampling event. The value wasn't readily accessible from outside the GC and @Maoni0 recommended that it probably isn't very useful so it is easiest to drop it. If you have any concerns let me know.

@chrisnas - do you want to give this branch a spin with your profiler to test it out? I'm guessing this is the final source or pretty close to it.

**Comment by chrisnas** [comment]
> @brianrob @chrisnas - I wanted to give a heads up that I removed the HeapIndex parameter on the AllocationSampling event. The value wasn't readily accessible from outside the GC and @Maoni0 recommended that it probably isn't very useful so it is easiest to drop it. If you have any concerns let me know.

Sounds good to me: I never found a usage for it  :^)

> @chrisnas - do you want to give this branch a spin with your profiler to test it out? I'm guessing this is the final source or pretty close to it.

I'll run my tests on monday - the rest of the week is off for me sorry

**Comment by noahfalk** [comment]
> I'll run my tests on monday - the rest of the week is off for me sorry

Sounds great! Its not a rush :)

**Comment by noahfalk** [comment]
/ba-g remaining issues are deadletters for which an issue isn't appropriate

**Comment by noahfalk** [comment]
@jkotas - Thanks for the earlier feedback. I think that has been covered now but let me know if you have any other feedback. Thanks!

File: src/coreclr/minipal/Unix/CMakeLists.txt
Comment (jkotas) [inline src/coreclr/minipal/Unix/CMakeLists.txt:4]: Should this be part of the minipal static library https://github.com/dotnet/runtime/blob/main/src/native/minipal/CMakeLists.txt instead this?

Comment (noahfalk) [reply src/coreclr/minipal/Unix/CMakeLists.txt:4]: yeah, probably. 👍

Comment (noahfalk) [reply src/coreclr/minipal/Unix/CMakeLists.txt]: Fixed

**Comment by noahfalk** [comment]
/ba-g remaining issues are deadletters for which an issue isn't appropriate

**Comment by noahfalk** [comment]
@chrisnas - just let me know whenever you've had a chance to test.
@jkotas - all the feedback so far should be addressed, let me know if there is anything else.

Thanks!

**Comment by jkotas** [comment]
I do not have any additional feedback. As I have mentioned in https://github.com/dotnet/runtime/pull/104955#pullrequestreview-2205426994, this should get a review from somebody familiar with tracing.

File: src/native/minipal/xoshiro128pp.h
Comment (am11) [inline src/native/minipal/xoshiro128pp.h:21]: We are adding doxygen-style docs to minipal APIs:
```suggestion
/**
 * Initialize the Xoshiro128++ random number generator state
 *
 * @param pState Pointer to the Xoshiro128++ state structure to be initialized.
 * @param seed Seed value to initialize the generator. Different seeds produce different random sequences.
 */
void minipal_xoshiro128pp_init(struct minipal_xoshiro128pp* pState, uint32_t seed);

/**
 * Generate the next random number in the Xoshiro128++ sequence
 *
 * @param pState Pointer to the Xoshiro128++ state structure.
 * @return The next 32-bit unsigned random number in the sequence.
 */
uint32_t minipal_xoshiro128pp_next(struct minipal_xoshiro128pp* pState);
```

**Comment by noahfalk** [comment]
> As I have mentioned in https://github.com/dotnet/runtime/pull/104955#pullrequestreview-2205426994, this should get a review from somebody familiar with tracing.

@brianrob - would you mind taking a look if you haven't already?

**Comment by chrisnas** [comment]
> @chrisnas - just let me know whenever you've had a chance to test.

I'm back from a conference in Belgium and I'm starting to work on it.
I first need to change our profiler to leverage the new event and then compare to the really allocated objects

File: src/tests/tracing/eventpipe/randomizedallocationsampling/manual/Allocate/AllocateArraysOfDoubles.cs
Comment (noahfalk) [reply src/tests/tracing/eventpipe/randomizedallocationsampling/manual/Allocate/AllocateArraysOfDoubles.cs]: fixed

**Comment by noahfalk** [comment]
Thanks all!

As a heads up I'll be out for extended period starting the week after next. I'm hoping that review will be finished this week and we can merge this. That gives me all of next week to respond if there is any unexpected post-merge regression discovered. I think we've got a reasonable amount of testing that the feature is working correctly already, but if you did find something @chrisnas in your additional testing we'd still have a long runway to get it fixed before .NET 10 ships next year.

**Review by brianrob — APPROVED** [review]
LGTM.  Thanks @noahfalk!

**Comment by noahfalk** [comment]
/ba-g Still getting deadletter issues on Chrome debugger tests, but it appears to be unrelated and pre-existing

**Comment by MichalStrehovsky** [comment]
I'm seeing the newly added test failing in native AOT outerloop in #109842.

```
/root/helix/work/correlation/nativeaottest.sh /root/helix/work/workitem/e/tracing/eventpipe/randomizedallocationsampling/allocationsampling/ allocationsampling.dll ''
  0.0s: ==TEST STARTING==
  0.0s: Started sending sentinel events...
  0.0s: Connecting to EventPipe...
  0.0s: Creating EventPipeEventSource...
  0.0s: EventPipeEventSource created
  0.0s: Dynamic.All callback registered
  0.0s: Running optional trace validator
  0.0s: Finished running optional trace validator
  0.0s: Starting stream processing...
  0.0s: Saw new provider 'Microsoft-Windows-DotNETRuntime'
  0.1s: Saw sentinel event
  0.1s: Stopped sending sentinel events
  0.1s: Starting event generating action...
  0.5s: Allocated 400000 instances...
  1.0s: Allocated 800000 instances...
  2.0s: Allocated 1200000 instances...
  3.5s: Allocated 1600000 instances...
  4.8s: 2000000 instances allocated
  4.8s: Stopping event generating action
  4.8s: Sending StopTracing command...
  4.9s: Finished StopTracing command
  4.9s: Saw new provider 'Microsoft-DotNETCore-EventPipe'
  4.9s: Stopping stream processing
  4.9s: Dropped 0 events
  4.9s: Reader task finished
  4.9s: Validating optional callback...
  4.9s: AllocationSampled counts validation
  4.9s: Nb events: 2763
  4.9s: Nb object128: 0
  4.9s: ==TEST FINISHED: FAILED!==
Xunit.Sdk.EqualException: Assert.Equal() Failure: Values differ
Expected: 100
Actual:   -1
   at Xunit.Assert.Equal[T](T, T, IEqualityComparer`1) + 0x24b
   at __GeneratedMainWrapper.Main() + 0x4f
```

**Comment by noahfalk** [comment]
> I'm seeing the newly added test failing in native AOT outerloop in https://github.com/dotnet/runtime/pull/109842.

@jkotas - gave me the heads up earlier today and filed https://github.com/dotnet/runtime/issues/109828. I'm investigating it today.

**Pull request merged**
